### PR TITLE
[BE] Password update

### PR DESCRIPTION
**Pull request by nspc97**
[BE] Password update
https://trello.com/c/fIH8X2Hw/41-be-password-update

Changed register and user_update logic.

### CODE REVIEW

**Review by MihaiDruta — APPROVED** [review]

File: src/Controller/UserController.php [hunk 1]
```diff
@@ -89,16 +90,16 @@ public function editUser($id, SerializerInterface $serializer, Request $request,
         $data = $request->getContent();
```
Comment (SpartakusMd) [inline src/Controller/UserController.php:91]: We forgot to add a check that you can edit only your profile. Please compare the logged in user with the user which is being edited.

File: src/Controller/UserController.php [hunk 2]
```diff
@@ -145,4 +147,59 @@ public function deleteUser($id, UserRepository $userRepository): JsonResponse
...
+     * @throws ORMException
+     * @throws OptimisticLockException
+     */
+    public function userChangePassword($id, UserRepository $userRepository, Request $request): JsonResponse
+    {
+        $user = $userRepository->find($id);
+        if (!$user) {
+            return new JsonResponse(['message' => 'The user was not found!'], Response::HTTP_NOT_FOUND);
+        }
+        $data = $request->getContent();
```
Comment (SpartakusMd) [inline src/Controller/UserController.php]: Add a check that you can change only your password. Please compare the logged in user with the user which is being edited.

File: src/DTO/UserDTO.php [hunk 3]
```diff
@@ -16,86 +20,150 @@ class UserDTO
...
+     * @Serializer\Type("string")
+     * @Assert\NotNull(groups={"PasswordEdit"})
+     * @Assert\EqualTo(propertyPath="newPassword",
+     *     message="Passwords do not match.",
+     *     groups={"PasswordEdit"}
+     *     )
+     * @Serializer\Expose()
+     * @Groups({"PasswordEdit"})
+     */
+    public $confirmNewPassword;
```
Comment (SpartakusMd) [inline src/DTO/UserDTO.php]: You can rename the fields so you don't have 2 confirm fields. On register you would have password, confirmPassword and on password change you would have oldPassword, password, confirmPassword. The validation rules will also be simpler in this case.

File: src/Exceptions/NotValidOldPassword.php [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+
+namespace App\Exceptions;
+
+class NotValidOldPassword extends \Exception
+{
+    public function __construct(string $message = '')
+    {
+        parent::__construct($message);
+    }
+}
```
Comment (SpartakusMd) [inline src/Exceptions/NotValidOldPassword.php]: Add new lines at the end of files.

File: src/Service/UserTransformer.php [hunk 5]
```diff
@@ -90,4 +104,33 @@ public function transform(
...
+        if (!$match) {
+            $passwordDoNotMatch = new NotValidOldPassword(
+                'Old password not valid.'
+            );
+            throw $passwordDoNotMatch;
+        }
+        $entity->setPassword($this->encoder->encodePassword($entity, $dto->newPassword));
+        return $entity;
+    }
+}
```
Comment (SpartakusMd) [inline src/Service/UserTransformer.php]: New line.

File: src/Service/UserTransformer.php [hunk 6]
```diff
@@ -90,4 +104,33 @@ public function transform(
...
+    /**
+     * @param UserDTO $dto
+     * @return User
+     * @throws NotValidOldPassword
+     * @throws EntityNotFound
+     */
+    public function changePasswordTransform(
+        UserDTO $dto
+    ): User {
+        $entity = $this->userRepo->find($dto->id);
```
Comment (SpartakusMd) [inline src/Service/UserTransformer.php]: On change password you should have only the passwords in the requests. The user id is sent in URL. Pass the $user object when calling `changePasswordTransform()`

**Review by SpartakusMd — CHANGES_REQUESTED** [review]

File: src/Controller/RegisterController.php [hunk 7]
```diff
@@ -57,7 +57,7 @@ public function createUser(SerializerInterface $serializer, Request $request, Us
         $data = $request->getContent();
 
         /** @var DeserializationContext $context */
-        $context = DeserializationContext::create();
+        $context = DeserializationContext::create()->setGroups(array('UserCreate'));
 
         $userDTO = $serializer->deserialize(
```
Comment (MihaiDruta) [inline src/Controller/RegisterController.php]: Inject the serializer in `__construct()` method like `validator` and `transformer`

File: src/Controller/UserController.php [hunk 8]
```diff
@@ -80,50 +81,55 @@ public function getUserDetails(User $user): Response
      * @throws ORMException
      * @throws OptimisticLockException
      */
-    public function editUser($id, Request $request, UserRepository $userRepository)
+    public function editUser($id, Request $request, UserRepository $userRepository, UserInterface $user)
```
Comment (SpartakusMd) [inline src/Controller/UserController.php]: I would use `$currentUser` or `$loggedUser` or `$authenticatedUser` as it better reflects what kind of user is this.

File: src/Controller/UserController.php [hunk 9]
```diff
@@ -80,50 +81,55 @@ public function getUserDetails(User $user): Response
...
     {
-        $user = $userRepository->find($id);
-        if (!$user) {
-            return new JsonResponse(['message' => 'The user was not found!'], Response::HTTP_NOT_FOUND);
+        $userUsername = $user->getUsername();
+        $userPage = $userRepository->find($id);
```
Comment (SpartakusMd) [inline src/Controller/UserController.php]: This is a user, not a page. Use `$user` or `$affectedUser`.

File: src/Controller/UserController.php [hunk 10]
```diff
@@ -80,50 +81,55 @@ public function getUserDetails(User $user): Response
...
+        if ($userPage === null) {
+            return new JsonResponse(['message' => 'Access denied!'], Response::HTTP_NOT_FOUND);
         }
 
-        $data = $request->getContent();
+        $userPage = $userRepository->find($id)->getUsername();
+        if ($userUsername === $userPage) {
```
Comment (SpartakusMd) [inline src/Controller/UserController.php]: Compare user ids, not usernames ..

Comment (SpartakusMd) [reply src/Controller/UserController.php]: It should be the reverse. If the ids (or usernames in your case) don't match, return the error JSON, as so, you won't have too much identation and you should alway fail / return fast from the functions.

File: src/Controller/UserController.php [hunk 11]
```diff
@@ -151,55 +157,66 @@ public function deleteUser($id, UserRepository $userRepository): JsonResponse
...
-    public function userChangePassword($id, UserRepository $userRepository, Request $request): JsonResponse
-    {
-        $user = $userRepository->find($id);
-        if (!$user) {
-            return new JsonResponse(['message' => 'The user was not found!'], Response::HTTP_NOT_FOUND);
+    public function userChangePassword(
+        $id,
+        UserRepository $userRepository,
+        Request $request,
+        UserInterface $user
```
Comment (SpartakusMd) [inline src/Controller/UserController.php]: Same here, rename `$user`

File: src/Service/UserTransformer.php [hunk 12]
```diff
@@ -40,39 +40,54 @@ public function __construct(
...
+    }
 
+    /**
+     * @param UserDTO $dto
+     * @return User
+     * @throws EntityNotFound
+     */
+    public function editTransform(
+        UserDTO $dto,
+        $id
```
Comment (SpartakusMd) [inline src/Service/UserTransformer.php]: Send the User object, not the id. You already searched the user in the database, there's no reason to do it twice.

File: src/Service/UserTransformer.php [hunk 13]
```diff
@@ -90,4 +105,34 @@ public function transform(
...
+
+    /**
+     * @param UserDTO $dto
+     * @return User
+     * @throws NotValidOldPassword
+     * @throws EntityNotFound
+     */
+    public function changePasswordTransform(
+        UserDTO $dto,
+        $id
```
Comment (SpartakusMd) [inline src/Service/UserTransformer.php]: Send the User object, not the id. You already searched the user in the database, there's no reason to do it twice.

**Review by SpartakusMd — CHANGES_REQUESTED** [review]

File: src/Controller/UserController.php [hunk 14]
```diff
@@ -70,45 +71,43 @@ public function getUserDetails(User $user): Response
...
-    public function editUser($id, SerializerInterface $serializer, Request $request, UserRepository $userRepository)
+    public function editUser(User $user, Request $request, UserRepository $userRepository)
     {
-        $user = $userRepository->find($id);
-        if (!$user) {
-            return new JsonResponse(['message' => 'The user was not found!'], Response::HTTP_NOT_FOUND);
+        $authenticatedUser = $this->getUser();
+
+        if ($authenticatedUser->getId() !== $user->getId()) {
+            return new JsonResponse(['message' => 'Access denied!'], Response::HTTP_BAD_REQUEST);
```
Comment (SpartakusMd) [inline src/Controller/UserController.php:88]: Actualy it should be `Response::HTTP_FORBIDDEN` as the current user cannot edit this user. You can fix it after the merge.

File: src/Controller/UserController.php [hunk 15]
```diff
@@ -145,4 +144,53 @@ public function deleteUser($id, UserRepository $userRepository): JsonResponse
...
+     */
+    public function userChangePassword(
+        User $user,
+        UserRepository $userRepository,
+        Request $request
+    ): JsonResponse {
+        $authenticatedUser = $this->getUser();
+
+        if ($authenticatedUser->getId() !== $user->getId()) {
+            return new JsonResponse(['message' => 'Access denied!'], Response::HTTP_BAD_REQUEST);
```
Comment (SpartakusMd) [inline src/Controller/UserController.php:166]: Actualy it should be Response::HTTP_FORBIDDEN as the current user cannot edit this user. You can fix it after the merge.

File: src/Controller/UserController.php [hunk 16]
```diff
@@ -145,4 +144,53 @@ public function deleteUser($id, UserRepository $userRepository): JsonResponse
...
+        }
+            $data = $request->getContent();
```
Comment (SpartakusMd) [inline src/Controller/UserController.php:168]: Indentation is wrong in this method :)

File: src/Exceptions/NotValidOldPassword.php [hunk 17]
```diff
@@ -0,0 +1,11 @@
+<?php
+
+namespace App\Exceptions;
+
+class NotValidOldPassword extends \Exception
+{
+    public function __construct(string $message = '')
```
Comment (SpartakusMd) [inline src/Exceptions/NotValidOldPassword.php:7]: Remove the construct if it doesn't do anything.

**Review by SpartakusMd — APPROVED** [review]

**Pull request merged**